cardano-ledger-test/src/Test/Cardano/Ledger/BaseTypes.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.BaseTypes where
+
+import qualified Cardano.Binary as Binary
+import Cardano.Ledger.BaseTypes
+import Data.Aeson
+import Data.ByteString.Lazy (ByteString)
+import Data.Either
+import Data.GenValidity (genValid)
+import Data.GenValidity.Scientific ()
+import Data.Scientific
+import Data.Typeable
+import Test.QuickCheck
+import Test.Shelley.Spec.Ledger.Serialisation.EraIndepGenerators ()
+import Test.Tasty
+import Test.Tasty.HUnit
+import Test.Tasty.QuickCheck
+
+expectLeft :: Either a b -> Assertion
+expectLeft = assertBool "Expected Left" . isLeft
+
+boundedRationalTests ::
+  forall a.
+  ( BoundedRational a,
+    ToJSON a,
+    FromJSON a,
+    Binary.ToCBOR a,
+    Binary.FromCBOR a,
+    Arbitrary a,
+    Show a,
+    Ord a
+  ) =>
+  [(String, ByteString)] ->
+  TestTree
+boundedRationalTests badJSONValues =
+  testGroup
+    (showsTypeRep (typeRep (Proxy :: Proxy a)) "")
+    [ testGroup
+        "Rational roundtrip"
+        [ testProperty "(boundRational . unboundRational)" $ \(bi :: a) ->
+            Just bi === boundRational (unboundRational bi),
+          testProperty "(unboundRational . boundRational)" $
+            forAll genValid $ \r ->
+              maybe
+                (property True)
+                ((r ===) . unboundRational)
+                (boundRational r :: Maybe a),
+          testProperty "bounding produces valid values within bounds" $
+            forAll genValid $ \r ->
+              case boundRational r of
+                Nothing -> property True
+                Just (br :: a) ->
+                  conjoin
+                    [ minBound <= br,
+                      br <= maxBound,
+                      unboundRational (minBound :: a) <= r,
+                      r <= unboundRational (maxBound :: a)
+                    ]
+        ],
+      testGroup "JSON" $
+        [ testProperty "ToJSON/FromJSON roundtrip up to an epsilon" $ \(br :: a) ->
+            within 500000 $
+              case eitherDecode (encode br) of
+                Left err -> error err
+                Right (br' :: a) ->
+                  epsilonMaybeEq 1e-18 (unboundRational br) (unboundRational br'),
+          testProperty "Roundtrip to Scientific and back up to an epsilon" $ \ui ->
+            within 500000 $
+              case fromRationalRepetendLimited 20 (unboundRational ui) of
```

Comment:
It is this crazy function: https://hackage.haskell.org/package/scientific-0.3.7.0/docs/Data-Scientific.html#v:fromRationalRepetendLimited that allows conversion from rational with guarantees that it will not loop forever.

According to google: 

> repetend -  the repeating figure or figures of a recurring decimal fraction.

Reply:
TIL!